### PR TITLE
ratelimit: allow metadata descriptors to be skipped

### PR DESCRIPTION
**Pull request by thomasvnoort**
Signed-off-by: Thomas van Noort <thomas.vannoort@datadoghq.com>

Commit Message: ratelimit: allow metadata descriptors to be skipped
Risk Level: low
Testing: added unit tests
Docs Changes: per the protobuf definitions
Release Notes: N/A
Platform Specific Features: N/A

Additional Description:
The default behaviour was to skip calling the rate limiting service whenever the metadata key was not found and the default value was not set. This was not correctly documented (nor tested) since it mentioned that only the descriptor was skipped whereas the rate limiting service was skipped altogether.

This adds a skip_if_absent field in the same spirit as for the request headers action: if set to true it skips the descriptor but still calls the rate limiting service, otherwise it skips the rate limiting service.

Note that the deprecated dynamic metadata action does not support this field and defaults to false.

### CODE REVIEW

**Comment by alyssawilk** [comment]
@kyessenov I'll let you throw it over to senior maintainers once you've signed off

File: test/common/router/router_ratelimit_test.cc
Comment (kyessenov) [inline test/common/router/router_ratelimit_test.cc:896]: nit: rename from `Empty2` to `EmptySkipIfAbsent`

**Review by kyessenov — COMMENTED** [review]
LGTM, need @envoyproxy/senior-maintainers and API review.

**Comment by kyessenov** [comment]
Throwing this over to API reviewers @markdroth . Changes to the xDS require a separate approval. Code LGTM.

**Comment by kyessenov** [comment]
/retest

Gentle ping @markdroth

**Comment by kyessenov** [comment]
I think this needs main marge to pick up CI changes.

**Review by kyessenov — APPROVED** [review]
Throwing over to @markdroth for API review.

**Comment by kyessenov** [comment]
CC @htuch 
Needs an API review.

**Review by htuch — COMMENTED** [review]
/lgtm api

File: test/common/router/router_ratelimit_test.cc
Comment (htuch) [inline test/common/router/router_ratelimit_test.cc:930]: It looks like there are two ways to configure this behavior based on the API changes, do these tests cover both?

Comment (kyessenov) [reply test/common/router/router_ratelimit_test.cc:930]: Yeah, I think coverage-wise, we only need to add one extra test with the flag on.

**Pull request merged**
